### PR TITLE
[GH-54] sync with latest mattermost-plugin-starter-template

### PR DESCRIPTION
**Pull request by jfrerich**
#### Summary
Syncs with latest mattermost-plugin-starter-template

#### Ticket Link
Fixes:
#54 
#69

### CODE REVIEW

**Comment by jfrerich** [comment]
I quite like having colorized terminal output and although I am pulling this OUT of the Makefile, what do you guys think about adding this color highlighting to our Makefile :) 

https://github.com/mattermost/mattermost-plugin-confluence/pull/68/files#diff-b67911656ef5d18c4ae36cb6741b7965L27-L34

**Comment by mickmister** [comment]
@jfrerich Do you mind going in detail with your process how you performed this sync? https://github.com/mattermost/mattermost-plugin-confluence/pull/68

**Comment by jfrerich** [comment]
@mickmister, I followed the steps found inside https://github.com/mattermost/mattermost-plugin-confluence/issues/69. 

It would be great to have this PR for doing this. I will have to get back to that ticket and see if I can repro much of the changes needed for the syncing in this PR.

**Comment by mickmister** [comment]
@jfrerich Any reason we can't use the plugin repo [sync tool](https://github.com/mattermost/mattermost-plugin-starter-template/pull/86)? Here's my test drive of using it https://github.com/mattermost/mattermost-plugin-todo/pull/103 , including the [plan.yml](https://github.com/mattermost/mattermost-plugin-todo/blob/f7f527b9ac4c4ef181b279aaa80e6b1f2840c020/build/sync/plan.yml) I used to sync. You can adjust the [original plan.yml](https://github.com/mattermost/mattermost-plugin-starter-template/blob/4ba779847102c8af3a3d53f8dc37ef5875bacfbe/build/sync/plan.yml) as you see fit.

**Comment by jfrerich** [comment]
Nice.  I hadn't seen the sync tool in use!  I'll have to test it out also and see how it reacts against my changes.

**Comment by jfrerich** [comment]
So, I did find some changes that I missed.  I also have quite a few differences with `.editorconfig` and `webapp/.eslintrc.json`.
* .circleci/config.yml` was drastically reduced.  
* `.editorconfig` updated
* `webapp/.eslintrc.json` updated

**Comment by jfrerich** [comment]
FYI, this is the command I am running: `go run ./build/sync/main.go ../mattermost-plugin-confluence/build/sync/plan.yml ../mattermost-plugin-confluence`
I updated the plan.yml file for this repo, but feel like I should have to do that..  There should be certain files that are always the same across repos.

File: .circleci/config.yml [hunk 1]
```diff
@@ -1,53 +1,12 @@
...
 workflows:
   version: 2
-  nightly:
-    triggers:
-      - schedule:
-          cron: "0 0 * * *"
-          filters:
-            branches:
-              only:
-                - master
```
Comment (hanzei) [inline .circleci/config.yml:15]: Why did this file go changed? It should use the version from https://github.com/mattermost/mattermost-plugin-demo/blob/master/.circleci/config.yml

File: build/.gitignore [hunk 2]
```diff
@@ -0,0 +1 @@
+bin
```
Comment (hanzei) [inline build/.gitignore:1]: Somehow `build/bin/deploy` got into this commit

File: build/pluginctl/main.go [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+	case "enable":
+		return enablePlugin(client, os.Args[2])
+	case "reset":
+		return resetPlugin(client, os.Args[2])
+	default:
+		return errors.New("invalid second argument")
+	}
+}
+
+func getClient() (*model.Client4, error) {
```
Comment (hanzei) [inline build/pluginctl/main.go:57]: https://github.com/mattermost/mattermost-plugin-starter-template/commit/4523a0a25eb0f34604cefac8a5e588d5450470f5 is quite helpful to have. Could you please also include it?

File: build/sync/plan.yml [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+      - webapp/tsconfig.json
+      - webapp/.eslintrc.json
+      # - webapp/src/index.js
+      # - server/configuration.go
+      # - server/plugin.go
+      # - server/plugin_test.go
+      # - public/hello.html
+      - .circleci/config.yml
+      # - webapp/i18n/en.json
+      # - webapp/package.json
```
Comment (hanzei) [inline build/sync/plan.yml:39]: Why not include `webapp/package.json`?

File: build/sync/plan.yml [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+        params:
+          create: true
+        # conditions:
+        #   - type: file_unaltered
+        #     params:
+        #       repo: plugin
+        #       reference-repo: template
+  - paths:
+      # - go.mod
+      # - webapp/webpack.config.js
```
Comment (hanzei) [inline build/sync/plan.yml:29]: Why not include `webapp/webpack.config.js`?

File: build/sync/plan.yml [hunk 6]
```diff
@@ -0,0 +1,51 @@
+checks:
+  - type: repo_is_clean
+    params:
+      repo: template
+  - type: repo_is_clean
+    params:
+      repo: plugin
+actions:
+  # - paths:
```
Comment (hanzei) [inline build/sync/plan.yml:9]: Nit: Mind removing the comments?

File: build/sync/plan.yml (hunk 4 above)
Comment (mickmister) [reply build/sync/plan.yml:39]: I recommend including `package.json` and `webpack.config.js` in the plan, then resolving diffs manually afterwards. This worked pretty well for me https://github.com/mattermost/mattermost-plugin-todo/blob/f7f527b9ac4c4ef181b279aaa80e6b1f2840c020/build/sync/plan.yml

File: .circleci/config.yml (hunk 1 above)
Comment (mickmister) [reply .circleci/config.yml:15]: Did you mean to link to the starter-template instead?

Comment (mickmister) [reply .circleci/config.yml:15]: nvm I see your comment [here](https://github.com/mattermost/mattermost-plugin-todo/pull/103#discussion_r462933898) explaining why :+1:

Comment (jfrerich) [reply .circleci/config.yml]: This is a point of confusion that we should address.  Thanks for the linking the comment regarding why we sync this file with demo instead of the starter template

File: build/sync/plan.yml (hunk 4 above)
Comment (jfrerich) [reply build/sync/plan.yml]: New commit syncs `webapp/package.json` and `webpack.config.js`

performed the following, post sync, to get these required packages.

`npm install --save superagent` 
`npm install --save js-cookie`

File: build/sync/plan.yml (hunk 6 above)
Comment (jfrerich) [reply build/sync/plan.yml]: Done

**Comment by jfrerich** [comment]
ok! There were quite a few commits needed after addressing those latest comments (syncing webpack, package.json, socket API), but this is ready for re-review

File: build/sync/plan.yml [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+  - paths:
+      - .circleci/config.yml
+      - webapp/.eslintrc.json
+      - webapp/package.json
+      - webapp/tsconfig.json
+      - webapp/webpack.config.js
+    actions:
+      - type: overwrite_file
+        params:
+          create: true
```
Comment (hanzei) [inline build/sync/plan.yml:25]: Should the two `path` get combined?

Comment (jfrerich) [reply build/sync/plan.yml:25]: Actually, yes they can.

File: server/service/get_subscription.go [hunk 8]
```diff
@@ -8,7 +8,7 @@ import (
 	"github.com/mattermost/mattermost-plugin-confluence/server/serializer"
 )
 
-const generalError = "Some error occurred. Please try again after sometime."
+const generalError = "some error occurred. Please try again after sometime"
```
Comment (mickmister) [inline server/service/get_subscription.go:11]: Just out of curiosity, why was this changed to lowercase? Same with other similar string changes

File: webapp/src/client/client.js [hunk 9]
```diff
@@ -9,10 +9,10 @@ import Constants from '../constants';
...
     constructor() {
         const url = new URL(window.location.href);
-        this.baseUrl = `${url.protocol}//${url.host}`;
-        this.pluginUrl = `${this.baseUrl}/plugins/${Constants.PLUGIN_NAME}`;
-        this.apiUrl = `${this.baseUrl}/api/v4`;
-        this.pluginApiUrl = `${this.pluginUrl}/api/v1`;
+        this.baseUrl = url.protocol + '//' + url.host;
+        this.pluginUrl = this.baseUrl + '/plugins/' + Constants.PLUGIN_NAME;
+        this.apiUrl = this.baseUrl + '/api/v4';
+        this.pluginApiUrl = this.pluginUrl + '/api/v1';
```
Comment (mickmister) [inline webapp/src/client/client.js:15]: Why were these changed to using the `+` syntax?

File: webapp/webpack.config.js [hunk 10]
```diff
@@ -10,52 +42,37 @@ module.exports = {
...
                 use: {
                     loader: 'babel-loader',
                     options: {
                         cacheDirectory: true,
-                        plugins: [
-                            '@babel/plugin-proposal-class-properties',
-                            '@babel/plugin-syntax-dynamic-import',
-                            '@babel/proposal-object-rest-spread',
-                        ],
-                        presets: [ // Babel configuration is in .babelrc because jest requires it to be there.
```
Comment (mickmister) [inline webapp/webpack.config.js:30]: Interesting how the old code here said the config was in a separate file, but then it went on to define/duplicate the config in this file.

File: webapp/src/constants/action_types.js [hunk 11]
```diff
@@ -1,7 +1,7 @@
 import {PLUGIN_NAME} from './manifest';
 
 export const ACTION_TYPES = {
-    OPEN_SUBSCRIPTION_MODAL: `${PLUGIN_NAME}_open_subscription_modal`,
-    CLOSE_SUBSCRIPTION_MODAL: `${PLUGIN_NAME}_close_subscription_modal`,
-    RECEIVED_SUBSCRIPTION: `${PLUGIN_NAME}_received_subscription`,
+    OPEN_SUBSCRIPTION_MODAL: PLUGIN_NAME + '_open_subscription_modal',
+    CLOSE_SUBSCRIPTION_MODAL: PLUGIN_NAME + '_close_subscription_modal',
+    RECEIVED_SUBSCRIPTION: PLUGIN_NAME + '_received_subscription',
```
Comment (mickmister) [inline webapp/src/constants/action_types.js:6]: Is something in the eslint config suggesting this change?

File: webapp/src/constants/action_types.js [hunk 12]
```diff
@@ -1,7 +1,7 @@
 import {PLUGIN_NAME} from './manifest';
```
Comment (mickmister) [inline webapp/src/constants/action_types.js:1]: `PLUGIN_NAME` is hardcoded in this imported manifest file. Can we instead have it derived from the generated manifest file? Non-blocking for approval

Comment (jfrerich) [reply webapp/src/constants/action_types.js:1]: I'm replacing `PLUGIN_NAME` from `constants/manifest.js` with `id` from `webapp/src/manifest.js`

File: webapp/src/client/client.js (hunk 9 above)
Comment (jfrerich) [reply webapp/src/client/client.js]: Leaving these with the `${url.protocol}` causes eslint to fail.  I don't like the change, but made it to move forward with the PR

File: webapp/src/constants/action_types.js (hunk 11 above)
Comment (jfrerich) [reply webapp/src/constants/action_types.js]: *Unfortunately* yes this caused eslint to fail.

**Comment by jfrerich** [comment]
@mickmister I went ahead and addressed some of your comments in another commit.  Can you please re-review?

File: webapp/src/constants/action_types.js (hunk 11 above)
Comment (mickmister) [reply webapp/src/constants/action_types.js]: What rules caused it to fail? The config should be updated to support the better syntax

Comment (jfrerich) [reply webapp/src/constants/action_types.js]: Apparently this is a bug  https://github.com/eslint/eslint/issues/13248

File: server/service/get_subscription.go (hunk 8 above)
Comment (mickmister) [reply server/service/get_subscription.go]: What about the grammar suggestions regarding capitalization? Is there a reason why this was made lowercase?

Comment (jfrerich) [reply server/service/get_subscription.go]: It was a linting error, so I changed it

Comment (mickmister) [reply server/service/get_subscription.go]: What was the linting error? It seems incorrect to me to display an error message that begins with a lowercase letter, when it is meant to be a sentence. Definitely something we want to fix in our lint settings rather than doing it this way 2/5
